common/app/concurrent/FutureSemaphore.scala

```diff
@@ -0,0 +1,19 @@
+package concurrent
+
+import java.util.concurrent.Semaphore
+import scala.concurrent.{ExecutionContext, Future}
+
+class FutureSemaphore(maxOperations: Int)(implicit ec: ExecutionContext) {
+  val semaphore = new Semaphore(maxOperations)
+
+  def execute[A](task: => Future[A]): Future[A] = {
+    if (semaphore.tryAcquire()) {
+      val result = task
+      result.foreach(_ => semaphore.release())
+      result.failed.foreach(_ => semaphore.release())
```

Comment:
Can using result.onComplete avoid the duplication here by handling both success and failure cases in one go?